src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(T item) {
+            this.item = item;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> temporaryItem = tail;
+        Node<T> newNode = new Node<>(value);
+        tail = newNode;
+        if (temporaryItem == null) {
+            head = newNode;
+        } else {
+            temporaryItem.next = newNode;
+            newNode.prev = temporaryItem;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> temporaryItem = head;
+            Node<T> newNode = new Node<>(value);
+            for (int i = 0; i <= index; i++) {
+                if (i == index) {
+                    if (temporaryItem == head) {
+                        temporaryItem.prev = newNode;
+                        newNode.next = temporaryItem;
+                        head = newNode;
+                    } else {
+                        temporaryItem.prev.next = newNode;
+                        newNode.prev = temporaryItem.prev;
+                        temporaryItem.prev = newNode;
+                        newNode.next = temporaryItem;
+                    }
+                    size++;
+                }
+                temporaryItem = temporaryItem.next;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            Node<T> temporaryItem = tail;
+            Node<T> newNode = new Node<>(list.get(i));
+            tail = newNode;
+            if (temporaryItem == null) {
+                head = newNode;
+            } else {
+                temporaryItem.next = newNode;
+                newNode.prev = temporaryItem;
+            }
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndex(index);
+        Node<T> temporaryItem = head;
+        for (int i = 0; i <= index; i++) {
+            if (i == index) {
+                return temporaryItem.item;
+            }
+            temporaryItem = temporaryItem.next;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        checkIndex(index);
+        Node<T> temporaryItem = head;
+        Node<T> newNode = new Node<>(value);
+        for (int i = 0; i <= index; i++) {
+            if (i == index) {
+                T element = temporaryItem.item;
+                temporaryItem.item = newNode.item;
+                return element;
+            }
+            temporaryItem = temporaryItem.next;
+        }
         return null;
```

Comment:
Instead of creating a new `Node` instance, you can directly set the `item` of the existing node. Also, there's no need for the loop since you can find the node directly using a method like `findNodeByIndex`.